File: Scala/0008.scala
object Solution {
  def myAtoi(s: String): Int = {
    val maxInt = 2147483647
    val minInt = -2147483648

    // Step 1: Ignore leading whitespace
    var index = 0
    var n = s.length
    while (index < n && s(index) == ' ') {
      index += 1
    }

    if (index == n) return 0  // If all spaces, return 0

    // Step 2: Check the sign
    var sign = 1
    if (index < n && (s(index) == '-' || s(index) == '+')) {
      if (s(index) == '-') sign = -1
      index += 1
    }

    // Step 3: Convert the number
    var result = 0
    while (index < n && s(index).isDigit) {
      val digit = s(index) - '0'
      // Check for overflow before multiplying and adding
      if (result > (maxInt - digit) / 10) {
        return if (sign == 1) maxInt else minInt
      }
      result = result * 10 + digit
      index += 1
    }

    sign * result
  }
}

File: C#/0008.cs
public class Solution {
    public int MyAtoi(string s) {
        if (string.IsNullOrWhiteSpace(s)) {
            return 0;
        }

        int index = 0;
        int n = s.Length;
        
        // Step 1: Ignore leading whitespaces
        while (index < n && s[index] == ' ') {
            index++;
        }
        
        // Step 2: Determine the sign
        int sign = 1;
        if (index < n && (s[index] == '-' || s[index] == '+')) {
            sign = (s[index] == '-') ? -1 : 1;
            index++;
        }
        
        // Step 3: Convert the number
        long result = 0;
        while (index < n && char.IsDigit(s[index])) {
            result = result * 10 + (s[index] - '0');
            index++;
            
            // Step 4: Check for overflow
            if (result * sign > int.MaxValue) return int.MaxValue;
            if (result * sign < int.MinValue) return int.MinValue;
        }
        
        // Return the final result
        return (int)(result * sign);
    }
}
